include/sparrow/array/external_array_data.hpp

```diff
@@ -0,0 +1,329 @@
+// Copyright 2024 Man Group Operations Limited
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#pragma once
+
+#include <cstring>
+#include <memory>
+#include <type_traits>
+
+#include "sparrow/c_interface.hpp"
+#include "sparrow/array/data_type.hpp"
+#include "sparrow/buffer/dynamic_bitset.hpp"
+#include "sparrow/buffer/buffer_view.hpp"
+#include "sparrow/utils/memory.hpp"
+
+namespace sparrow
+{
+    namespace impl
+    {
+        template <class T>
+        class external_wrapper
+        {
+        public:
+
+            using self_type = external_wrapper<T>;
+
+            template <class U>
+            requires std::is_rvalue_reference_v<U&&>
+            external_wrapper(U&&, bool own);
+
+            template <class U>
+            requires std::is_pointer_v<U>
+            external_wrapper(U, bool own);
+
+            T& data();
+            const T& data() const;
+
+        private:
+
+            struct no_releaser
+            {
+                void operator()(T*) const {}
+            };
+
+            struct releaser
+            {
+                void operator()(T* t) const;
+            };
+
+            std::shared_ptr<T> p_data = nullptr;
```

Comment:
Yes, because we don't know how to allocate and copy the  ArrowSchema and ArrowArray yet. This is something we could improve, but that requires more involved development.